contrib/internal/httptrace/httptrace.go

```diff
@@ -26,22 +26,54 @@ var (
 	cfg = newConfig()
 )
 
+type inferredSpanCreatedCtxKey struct{}
+
+type FinishSpanFunc = func(status int, errorFn func(int) bool, opts ...tracer.FinishOption)
+
 // StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags (http.method, http.url,
 // http.useragent). Any further span start option can be added with opts.
-func StartRequestSpan(r *http.Request, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context) {
+func StartRequestSpan(r *http.Request, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context, FinishSpanFunc) {
 	// Append our span options before the given ones so that the caller can "overwrite" them.
 	// TODO(): rework span start option handling (https://github.com/DataDog/dd-trace-go/issues/1352)
 
 	var ipTags map[string]string
 	if cfg.traceClientIP {
 		ipTags, _ = httpsec.ClientIPTags(r.Header, true, r.RemoteAddr)
 	}
+
 	nopts := make([]ddtrace.StartSpanOption, 0, len(opts)+1+len(ipTags))
+	inferredStartSpanOpts := make([]ddtrace.StartSpanOption, 0, 1)
+
+	spanParentCtx, spanParentErr := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header))
+	var spanLinksCtx ddtrace.SpanContextWithLinks
+
+	if spanParentErr == nil {
+		if linksCtx, ok := spanParentCtx.(ddtrace.SpanContextWithLinks); ok {
+			spanLinksCtx = linksCtx
+			inferredStartSpanOpts = append(inferredStartSpanOpts, tracer.WithSpanLinks(spanLinksCtx.SpanLinks()))
+		}
+	}
```

Comment:
@zarirhamza @jordan-wong Given that `inferredStartSpanOpts` is only used in case the condition `cfg.inferredProxyServicesEnabled && !inferredProxySpanCreated` is true, I'd move  this block, and the declaration for the variable into that if block.

It would remove one allocation in the hot path, as the benchmarks are reporting.